src/Sentry.Tunnel/SentryTunnelMiddleware.cs

```diff
@@ -0,0 +1,85 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Reflection;
+using System.Text.Json;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Net.Http.Headers;
+
+namespace Sentry.Tunnel
+{
+    public class SentryTunnelMiddleware : IMiddleware
+    {
+        private readonly string[] _allowedHosts;
+
+        public SentryTunnelMiddleware(string[] allowedHosts)
+        {
+            _allowedHosts = new[] {"sentry.io"}.Concat(allowedHosts).ToArray();
+        }
+
+        public async Task InvokeAsync(HttpContext context, RequestDelegate next)
+        {
+            if (context.Request.Method == "OPTIONS")
+            {
+                context.Response.Headers.Add("Access-Control-Allow-Origin", new[] {(string) context.Request.Headers["Origin"]});
+                context.Response.Headers.Add("Access-Control-Allow-Headers", new[] {"Origin, X-Requested-With, Content-Type, Accept"});
+                context.Response.Headers.Add("Access-Control-Allow-Methods", new[] {"POST, OPTIONS"});
+                context.Response.Headers.Add("Access-Control-Allow-Credentials", new[] {"true"});
+                context.Response.StatusCode = 200;
+                return;
+            }
+
+            var httpClientFactory = context.RequestServices.GetRequiredService<IHttpClientFactory>();
+            var client = httpClientFactory.CreateClient("SentryTunnel");
+            client.DefaultRequestHeaders.UserAgent.Add(new ProductInfoHeaderValue("Sentry.NET Tunnel", GetType().Assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion));
```

Comment:
I'm not sure if the whole `ProductInfoHeaderValue` object can be reused. If so, we could just hold a reference as a field otherwise we could just cache `GetType().Assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion)`